### PR TITLE
Add soft length regulation for sequence generation

### PR DESCRIPTION
**Pull request by kevinpl07**
# What does this PR do?

This PR enables users to softly regulate the length when using sampling method in model.generate(). We had the use case where we wanted to keep a generated sequence adequately short with no "hard" cutoff - we wanted the model to come to an end without just setting max_length, so that the end_token is also predicted at a meaningful position.

The SoftLengthLogitsProcessor exponentially increases the score of the eos_token_id until it is generated. It can be configured by these parameters inside model.generate():
- length_regulation_start
- length_regulation_factor

I am aware that there is already a length_penalty in place, but as far as I understood it is used in beam search and did not fulfil our need.

Usage:
```python
gen_tokens = model.generate(input_ids, do_sample=True, temperature=0.9, max_length=600, length_regulation_start=300, length_regulation_factor=1.01)
```

Example:
Here are the lengths of 5 samples generated for a length-regulated and default sequence-generation. The goal was to end the sequence between 300 and 400 tokens without a hard cutoff.
Code
```python
print("Regular output with max_length=600")
for index in range(5):
    gen_tokens = model.generate(input_ids, do_sample=True, temperature=0.9, max_length=600, pad_token_id=50256)
    print(gen_tokens.shape[1])

print("--------------")
print("Soft regulation (length_regulation_start=300, length_regulation_factor=1.01):")
for index in range(5):
    gen_tokens = model.generate(input_ids, do_sample=True, temperature=0.9, max_length=600, length_regulation_start=300, length_regulation_factor=1.01, pad_token_id=50256)
    print(gen_tokens.shape[1])
```
Results
```
Regular output with max_length=600
600
600
600
600
600
--------------
Soft regulation (length_regulation_start=300, length_regulation_factor=1.01)
361
365
381
370
396
```

Here we see one length-regulated example output. The <endoftext> token is being reasonably placed:
```
In recent events the conflict between Elon Musk and Donald Trump heated up. Trump criticised Musk after he was quoted by The New York Times saying that a ‘solution’ to the US’s problems in the technology industry might be “a bullet to the head”.

In return Musk sent the president a series of tweets where he called the president “a traitor”, and “an idiot” as well as said he wouldn’t take the challenge of building the president a $10 million Tesla Roadster.

“It’s not my style to engage in such juvenile behaviour — it’s never been my style to engage in such juvenile behaviour — but the stakes for positive change have become that important,” said Musk in an interview with Forbes magazine, which is to be published on Wednesday.

The interview was published a few days before Musk went public about buying the Boring Company, which he described in the interview as a “vertical tunnel boring machine.”

After a few hours after this, Musk issued a statement about his tweet regarding the Roadster and the $10 million challenge to the president.

“The idea that the government could take ownership of all privately owned infrastructure in the U.S., just because I want it, is absurd on its face.”

He said he was being “clearly sarcastic.”

This was enough for Trump and his advisors to get angry, with Trump telling CNN that Musk is “a very, very, different kind of person than me.”

“I just think that he’s very, very smart,” Trump added.<|endoftext|>
```

This works as well with other EOS tokens that may be used in few shot learning scenarios such as "\n".

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

**Comment by kevinpl07** [comment]
@patrickvonplaten, @LysandreJik

**Comment by patrickvonplaten** [comment]
Hey @kevinpl07,

Thanks a lot for your PR! Do you know whether this method is used in any paper? Could you give me more on its scientific background?

In general, I think it's ok to add such a method here though

File: src/transformers/generation_logits_process.py [hunk 1]
```diff
@@ -614,3 +614,31 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         scores[scores == float("inf")] = torch.finfo(scores.dtype).max
 
         return scores
+
+
+class SoftLengthLogitsProcessor(LogitsProcessor):
```
Comment (patrickvonplaten) [inline src/transformers/generation_logits_process.py:619]: Not a huge fan of the name - it is in the end some kind of length penalty to me so could we maybe change it to something like:

```
ExponentialDecayLengthPenalty
```
?

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -708,6 +717,10 @@ def _get_logits_processor(
             processors.append(ForcedEOSTokenLogitsProcessor(max_length, forced_eos_token_id))
         if remove_invalid_values is True:
             processors.append(InfNanRemoveLogitsProcessor())
+        if length_regulation_factor is not None and length_regulation_start is not None:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:720]: Think it would be cleaner to just have a single input tuple here `exponential_decay_length_penalty`. What do you think @Narsil ?

Comment (Narsil) [reply src/transformers/generation_utils.py:720]: I do agree a tuple is better.

The reason is if a user adds only 1 argument, then the current code would fail (since one would be None and the other have a real value).
So having a tuple forces users to set both at once, making it less likely to crash/fail.

Another thing though, is that this doesn't take into account already existing tokens (either in encode-decoder or in decoder-only). IMHO, this makes this argument harder to use, since you have to offset it with the start_ids, which you might not have access to (for instance pipeline users). `max_length` as the same issue and `max_new_tokens` is there to be agnostic to the initial length. 

If we could make `length_regulation_start` become expressed in generated tokens instead of overall tokens ,it would make a nicer api IMO.

File: src/transformers/generation_logits_process.py (hunk 1 above)
Comment (kevinpl07) [reply src/transformers/generation_logits_process.py:619]: You won't believe how i struggled finding a proper name for the logitsprocessor. I like ExponentialDecayLengthPenalty, will change it accordingly.

File: src/transformers/generation_utils.py (hunk 2 above)
Comment (kevinpl07) [reply src/transformers/generation_utils.py:720]: Tuple is fine for me. any ideas on an expressive name for the tuple? 

The pipeline arguments makes sense to me, I would change it in the same fashion as max_new_tokens. Do you agree with this change @patrickvonplaten?

**Comment by kevinpl07** [comment]
@patrickvonplaten 
Regarding scientific background of this method. We honestly just used intuition to solve our problem. It could very well be that this kind of LengthDecay is used in some paper.

File: src/transformers/generation_utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:720]: `config` is allowed to have tuples - tuples are immutable so that's totally fine for configs in my opinion. I'm also to have length regulation more dependent on the number of generated tokens as suggested by @Narsil! `max_new_tokens` is better than `max_length` in this regard. 

Regarding the name, I don't really mind that much as long as its docstring is good and the name closely resembles the logit processor class name.

**Comment by patrickvonplaten** [comment]
Hey @kevinpl07,

Cool think we can move forward then with this logits processor. It would also be nice to add a test in the end :-)

**Comment by patrickvonplaten** [comment]
@kevinpl07 - sorry I think the github commit history is broken. Could you resubmit a PR here? :-)

**Comment by kevinpl07** [comment]
> @kevinpl07 - sorry I think the github commit history is broken. Could you resubmit a PR here? :-)

@patrickvonplaten I updated my PR, with the agreed changes and tests. I don't know if that fixed the commit history problem. Will squashing the commit upon merging fix this? Otherwise let me know how I can clean the history for you.

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -1106,6 +1120,12 @@ def generate(
             # if decoder-only then inputs_tensor has to be `input_ids`
             input_ids = inputs_tensor
 
+        # Prepare exponential_decay_length_penalty, so decay start is applied to newly generated tokens
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1123]: Sorry, final change request. We are quite strict on not adding code that is to specific here. Could you instead add this code to `ExponentialDecayLengthPenalty` by passing the input_ids_shape as a parameter to init? 

This way all code specific to `ExponentialDecayLengthPenalty` is contained in this class

Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:1123]: Apart from this, the PR looks good to me :-)

Comment (kevinpl07) [reply src/transformers/generation_utils.py]: @patrickvonplaten sorry for the radio silence (was on vacation). I updated the PR :)

Comment (kevinpl07) [reply src/transformers/generation_utils.py]: @patrickvonplaten ping :)

File: src/transformers/generation_logits_process.py [hunk 4]
```diff
@@ -647,3 +647,31 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
         return scores
+
+
+class ExponentialDecayLengthPenalty(LogitsProcessor):
+    r"""
+    [`LogitsProcessor`] that exponentially increases the score of the eos_token_id after regulation_start has been
+    reached.
+
+    Args:
+        regulation_start (`int`):
```
Comment (patrickvonplaten) [inline src/transformers/generation_logits_process.py:658]: This the args need to be updated here no (`exponential_decay_length_penalty`)

File: src/transformers/generation_utils.py [hunk 5]
```diff
@@ -964,6 +975,11 @@ def generate(
                 crash. Note that using `remove_invalid_values` can slow down generation.
             synced_gpus (`bool`, *optional*, defaults to `False`):
                 Whether to continue running the while loop until max_length (needed for ZeRO stage 3)
+            exponential_decay_length_penalty (`Tuple`, *optional*):
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:978]:
```suggestion
            exponential_decay_length_penalty (`tuple(int, float)`, *optional*):
```

File: src/transformers/generation_utils.py [hunk 6]
```diff
@@ -843,6 +853,7 @@ def generate(
         forced_eos_token_id: Optional[int] = None,
         remove_invalid_values: Optional[bool] = None,
         synced_gpus: Optional[bool] = False,
+        exponential_decay_length_penalty: Optional[Tuple] = None,
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:856]:
```suggestion
        exponential_decay_length_penalty: Optional[Tuple[Union[int, float]]] = None,
```

**Review by patrickvonplaten — APPROVED** [review]
Looks very nice! Left some final nits (docstring), then we'll need to rebase the PR and we should be good to go :-)

File: src/transformers/generation_utils.py [hunk 7]
```diff
@@ -964,6 +975,11 @@ def generate(
                 crash. Note that using `remove_invalid_values` can slow down generation.
             synced_gpus (`bool`, *optional*, defaults to `False`):
                 Whether to continue running the while loop until max_length (needed for ZeRO stage 3)
+            exponential_decay_length_penalty (`Tuple`, *optional*):
+                This Tuple adds an exponentially increasing length penalty, after a certain amount of tokens have been
+                generated. The tuple shall consist of: (start_index, decay_factor) where start_index indicates where
+                penalty starts and decay_factor represents the factor of exponential decay
```
Comment (sgugger) [inline src/transformers/generation_utils.py:981]:
```suggestion
                generated. The tuple shall consist of: `(start_index, decay_factor)` where `start_index` indicates where
                penalty starts and `decay_factor` represents the factor of exponential decay
```

**Review by sgugger — APPROVED** [review]
LGTM! Thanks for adding this!

**Comment by patrickvonplaten** [comment]
Hey @kevinpl07 - could you maybe rebase your branch and I think then we can merge :-)

**Comment by Narsil** [comment]
@kevinpl07 ,

Be careful I think you added many unrelated files (687 were modified.).

If you're unsure how to fix it, tell us, we can probably help.

**Comment by kevinpl07** [comment]
@patrickvonplaten changed the docstrings accordingly and rebased :)

File: src/transformers/generation_utils.py [hunk 8]
```diff
@@ -653,6 +654,7 @@ def _get_logits_processor(
         repetition_penalty: float,
         no_repeat_ngram_size: int,
         encoder_no_repeat_ngram_size: int,
+        input_ids: torch.LongTensor,
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:657]: Sorry to only notice this now. I think we can make this clearer here. We don't actually need the `input_ids`, but all we need is the seq_length of the input_ids here. Also it would help the reader here to know that it corresponds to the sequence_length of the input to the decoder (in case it's an encoder-decoder model).

Could you maybe instead create a variable `input_ids_seq_length` and pass this one instead?

File: src/transformers/generation_utils.py [hunk 9]
```diff
@@ -1179,6 +1195,7 @@ def generate(
             repetition_penalty=repetition_penalty,
             no_repeat_ngram_size=no_repeat_ngram_size,
             encoder_no_repeat_ngram_size=encoder_no_repeat_ngram_size,
+            input_ids=input_ids,
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1198]: Can we replace all occurences af `input_ids.shape[-1]` with a variable called `input_ids_seq_length` and pass this one here instead?

**Comment by patrickvonplaten** [comment]
Don't worry about the `run_tests_hub` test - not related to this PR ;-)

**Comment by kevinpl07** [comment]
> 

So everything good from my side?

**Comment by kevinpl07** [comment]
@sgugger anything from my side, that I can do support the merge?

**Comment by patrickvonplaten** [comment]
Merging - thanks for your work here!
